Cannot modify benchmark classes between jmh runs without killing daemon or using --no-daemon

This bug is fully demonstrated, and described in more detail in this repository which is specifically for demonstrating it: https://github.com/ian-bertolacci/jmh-update-bug

(excerpts from the readme)

When developing, I sometimes make lots of changes until I like what I have.

In JMH, using the jmh-gradle-plugin (https://github.com/melix/jmh-gradle-plugin), this causes an issue, but only when using the gradle daemon.

To be specific, if I run `gradle clean build jmh`, then later change a class that JMH uses as a benchmark source, then run `gradle clean build jmh` again, this causes a `java.lang.IncompatibleClassChangeError` exception to be thrown

In the project, there is a benchmark definition source file `project/src/jmh/MyClassBenchmark.java`
It has an internal factory class, `FactoryA`.

This is fine, but lets say I then add a different factory, `FactoryB`, which is perfectly valid Java code and is perfectly acceptable to JMH (ignoring developer taste or best practices).

If working from a *clean* project, making the changes, and *then* running `gradle clean build jmh` is perfectly fine, and the project will build and the benchmark will run successfully.

In contrast, if working from a *clean* project, *first* running `gradle clean build jmh`, *then* making the changes, and *finally* running `gradle clean build jmh` again, it will throw the java.lang.IncompatibleClassChangeError.

This can be worked around by running gradle with `--no-daemon`, or by restarting *all* gradle daemons.

The full output for this case (and can be found in `outputs/project_updated_after_gradle_invocation_broken.2.out`):
```
> Task :clean
> Task :compileJava
> Task :processResources NO-SOURCE
> Task :classes
> Task :jar
> Task :assemble
> Task :compileTestJava NO-SOURCE
> Task :processTestResources NO-SOURCE
> Task :testClasses UP-TO-DATE
> Task :test NO-SOURCE
> Task :check UP-TO-DATE
> Task :build
> Task :compileJmhJava
> Task :processJmhResources NO-SOURCE
> Task :jmhClasses

> Task :jmhRunBytecodeGenerator FAILED
Processing 4 classes from /home/ian/Projects/TestCode/jmh-gradle-plugin-change-bug/project_updated_after_gradle_invocation_broken/build/classes/java/jmh with "reflection" generator
Writing out Java source to /home/ian/Projects/TestCode/jmh-gradle-plugin-change-bug/project_updated_after_gradle_invocation_broken/build/jmh-generated-sources and resources to /home/ian/Projects/TestCode/jmh-gradle-plugin-change-bug/project_updated_after_gradle_invocation_broken/build/jmh-generated-resources

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':jmhRunBytecodeGenerator'.
> A failure occurred while executing me.champeau.gradle.JmhBytecodeGeneratorRunnable
   > Generation of JMH bytecode failed with 1 errors:
       - Annotation generator had thrown the exception.
     java.lang.IncompatibleClassChangeError: com.cool.library.benchmarks.MyClassBenchmark and com.cool.library.benchmarks.MyClassBenchmark$FactoryB disagree on InnerClasses attribute
       at java.base/java.lang.Class.getDeclaringClass0(Native Method)
       at java.base/java.lang.Class.getEnclosingClass(Class.java:1517)
       at java.base/java.lang.Class.getCanonicalName0(Class.java:1625)
       at java.base/java.lang.Class.getCanonicalName(Class.java:1610)
       at org.openjdk.jmh.generators.reflection.RFClassInfo.getQualifiedName(RFClassInfo.java:67)
       at org.openjdk.jmh.generators.core.BenchmarkGenerator.buildAnnotatedSet(BenchmarkGenerator.java:206)
       at org.openjdk.jmh.generators.core.BenchmarkGenerator.generate(BenchmarkGenerator.java:75)
       at me.champeau.gradle.JmhBytecodeGeneratorRunnable.run(JmhBytecodeGeneratorRunnable.java:126)
       at org.gradle.workers.internal.AdapterWorkAction.execute(AdapterWorkAction.java:57)
       at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
       at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
       at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
       at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:97)
       at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
       at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:49)
       at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:30)
       at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:85)
       at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:55)
       at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
       at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
       at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
       at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
       at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
       at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
       at java.base/java.lang.reflect.Method.invoke(Method.java:566)
       at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
       at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
       at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
       at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
       at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
       at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
       at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
       at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
       at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
       at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
       at java.base/java.lang.Thread.run(Thread.java:834)
```

--- a/src/main/groovy/me/champeau/gradle/IsolatedRunner.java
+++ b/src/main/groovy/me/champeau/gradle/IsolatedRunner.java
@@ -16,31 +16,39 @@
 package me.champeau.gradle;
 
 import org.gradle.api.GradleException;
+import org.openjdk.jmh.runner.BenchmarkList;
+import org.openjdk.jmh.runner.CompilerHints;
 import org.openjdk.jmh.runner.Runner;
 import org.openjdk.jmh.runner.RunnerException;
 import org.openjdk.jmh.runner.options.Options;
-import org.openjdk.jmh.runner.options.OptionsBuilder;
 
-import javax.annotation.Nonnull;
 import javax.inject.Inject;
 import java.io.File;
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
 import java.util.Set;
 
 public class IsolatedRunner implements Runnable {
 
     private final Options options;
-    private Set<File> classpathUnderTest;
+    private final Set<File> classpathUnderTest;
+    private final File benchmarkList;
+    private final File compilerHints;
 
     @Inject
-    public IsolatedRunner(final Options options, final Set<File> classpathUnderTest) {
+    public IsolatedRunner(final Options options, final Set<File> classpathUnderTest, File benchmarkList, File compilerHints) {
         this.options = options;
         this.classpathUnderTest = classpathUnderTest;
+        this.benchmarkList = benchmarkList;
+        this.compilerHints = compilerHints;
     }
 
     @Override
     public void run() {
         String originalClasspath = System.getProperty("java.class.path");
         Runner runner = new Runner(options);
+        updateBenchmarkList(runner);
+        updateCompilerHints();
         try {
             System.setProperty("java.class.path", toPath(classpathUnderTest));
             // JMH uses the system property java.class.path to derive the runtime classpath of the forked JVM
@@ -55,6 +63,34 @@ public class IsolatedRunner implements Runnable {
                 System.clearProperty("java.class.path");
             }
         }
+    }
+
+    private void updateBenchmarkList(Runner runner) {
+        BenchmarkList benchmarkList = BenchmarkList.fromFile(this.benchmarkList.getAbsolutePath());
+        try {
+            Field listField = runner.getClass().getDeclaredField("list");
+            makeWriteable(listField);
+            listField.set(runner, benchmarkList);
+        } catch (Exception e) {
+            throw new GradleException("Error instantiating benchmarks", e);
+        }
+    }
+    private void updateCompilerHints() {
+        CompilerHints compilerHints = CompilerHints.fromFile(this.compilerHints.getAbsolutePath());
+        try {
+            Field listField = CompilerHints.class.getDeclaredField("defaultList");
+            makeWriteable(listField);
+            listField.set(null, compilerHints);
+        } catch (Exception e) {
+            throw new GradleException("Error instantiating benchmarks", e);
+        }
+    }
+
+    private void makeWriteable(Field listField) throws NoSuchFieldException, IllegalAccessException {
+        listField.setAccessible(true);
+        Field modifiersField = Field.class.getDeclaredField("modifiers");
+        modifiersField.setAccessible(true);
+        modifiersField.setInt(listField, listField.getModifiers() & ~Modifier.FINAL);
     }
 
     private String toPath(Set<File> classpathUnderTest) {

--- a/src/main/groovy/me/champeau/gradle/JMHPlugin.groovy
+++ b/src/main/groovy/me/champeau/gradle/JMHPlugin.groovy
@@ -89,6 +89,8 @@ class JMHPlugin implements Plugin<Project> {
         createTask(project, JMH_NAME, JMHTask) {
             it.group JMH_GROUP
             it.dependsOn project.jmhJar
+            it.benchmarkList = new File(jmhGeneratedResourcesDir, "META-INF/BenchmarkList")
+            it.compilerHints = new File(jmhGeneratedResourcesDir, "/META-INF/CompilerHints")
         }
 
         configureIDESupport(project)

--- a/src/main/groovy/me/champeau/gradle/JMHTask.java
+++ b/src/main/groovy/me/champeau/gradle/JMHTask.java
@@ -27,6 +27,7 @@ import org.gradle.workers.WorkerExecutor;
 import org.openjdk.jmh.runner.options.Options;
 
 import javax.inject.Inject;
+import java.io.File;
 
 /**
  * The JMH task converts our {@link JMHPluginExtension extension configuration} into JMH specific
@@ -39,6 +40,9 @@ public class JMHTask extends DefaultTask {
 
     private final WorkerExecutor workerExecutor;
 
+    private File benchmarkList;
+    private File compilerHints;
+
     @Inject
     public JMHTask(final WorkerExecutor workerExecutor) {
         this.workerExecutor = workerExecutor;
@@ -48,21 +52,18 @@ public class JMHTask extends DefaultTask {
     public void before() {
         final JMHPluginExtension extension = getProject().getExtensions().getByType(JMHPluginExtension.class);
         final Options options = extension.resolveArgs();
+
         extension.getResultsFile().getParentFile().mkdirs();
 
         workerExecutor.submit(IsolatedRunner.class, workerConfiguration -> {
             workerConfiguration.setIsolationMode(IsolationMode.PROCESS);
             ConfigurationContainer configurations = getProject().getConfigurations();
-            FileCollection classpath = configurations.getByName("jmh").plus(getProject().files(getJarArchive()));
+            workerConfiguration.classpath(configurations.getByName("jmh"));
+            FileCollection benchmarkClasspath = configurations.getByName("jmh").plus(getProject().files(getJarArchive()));
             if (extension.isIncludeTests()) {
-                classpath = classpath.plus(configurations.getByName("testRuntimeClasspath"));
+                benchmarkClasspath = benchmarkClasspath.plus(configurations.getByName("testRuntimeClasspath"));
             }
-            // TODO: This isn't quite right.  JMH is already a part of the worker classpath,
-            // but we need it to be part of the "classpath under test" too.
-            // We only need the jar for the benchmarks on the classpath so that the BenchmarkList resource reader
-            // can find the BenchmarkList file in the jar.
-            workerConfiguration.classpath(classpath);
-            workerConfiguration.params(options, classpath.getFiles());
+            workerConfiguration.params(options, benchmarkClasspath.getFiles(), benchmarkList, compilerHints);
             workerConfiguration.getForkOptions().getSystemProperties().put(JAVA_IO_TMPDIR, getTemporaryDir());
         });
     }
@@ -76,4 +77,11 @@ public class JMHTask extends DefaultTask {
         return ((Jar)getProject().getTasks().getByName(JMHPlugin.JMH_JAR_TASK_NAME)).getArchiveFile();
     }
 
+    public void setBenchmarkList(File benchmarkList) {
+        this.benchmarkList = benchmarkList;
+    }
+
+    public void setCompilerHints(File compilerHints) {
+        this.compilerHints = compilerHints;
+    }
 }

--- a/src/main/groovy/me/champeau/gradle/JmhBytecodeGeneratorRunnable.java
+++ b/src/main/groovy/me/champeau/gradle/JmhBytecodeGeneratorRunnable.java
@@ -36,16 +36,19 @@ import java.util.Map;
 import static org.openjdk.jmh.generators.bytecode.JmhBytecodeGenerator.*;
 
 public class JmhBytecodeGeneratorRunnable implements Runnable {
+    private final File[] classpath;
     private final File[] compiledBytecodeDirectories;
     private final File outputSourceDirectory;
     private final File outputResourceDirectory;
     private final String generatorType;
 
     @Inject
-    public JmhBytecodeGeneratorRunnable(final File[] compiledBytecodeDirectories,
+    public JmhBytecodeGeneratorRunnable(final File[] classpath,
+                                        final File[] compiledBytecodeDirectories,
                                         final File outputSourceDirectory,
                                         final File outputResourceDirectory,
                                         final String generatorType) {
+        this.classpath = classpath;
         this.compiledBytecodeDirectories = compiledBytecodeDirectories;
         this.outputSourceDirectory = outputSourceDirectory;
         this.outputResourceDirectory = outputResourceDirectory;
@@ -63,25 +66,27 @@ public class JmhBytecodeGeneratorRunnable implements Runnable {
             generatorType = DEFAULT_GENERATOR_TYPE;
         }
 
-        URL[] urls = new URL[compiledBytecodeDirectories.length];
+        URL[] urls = new URL[classpath.length + compiledBytecodeDirectories.length];
+        for (int i = 0; i < classpath.length; i++) {
+            try {
+                urls[i] = classpath[i].toURI().toURL();
+            } catch (MalformedURLException e) {
+                throw new RuntimeException(e);
+            }
+        }
         for (int i = 0; i < compiledBytecodeDirectories.length; i++) {
             try {
-                urls[i] = compiledBytecodeDirectories[i].toURI().toURL();
+                urls[i + classpath.length] = compiledBytecodeDirectories[i].toURI().toURL();
             } catch (MalformedURLException e) {
                 throw new RuntimeException(e);
             }
         }
 
-        // Include compiled bytecode on classpath, in case we need to
-        // resolve the cross-class dependencies
-        URLClassLoader amendedCL = new URLClassLoader(
-                urls,
-                this.getClass().getClassLoader());
-
+        URLClassLoader loader = new URLClassLoader(urls, this.getClass().getClassLoader());
         ClassLoader ocl = Thread.currentThread().getContextClassLoader();
 
         try {
-            Thread.currentThread().setContextClassLoader(amendedCL);
+            Thread.currentThread().setContextClassLoader(loader);
 
             FileSystemDestination destination = new FileSystemDestination(outputResourceDirectory, outputSourceDirectory);
 
@@ -113,7 +118,7 @@ public class JmhBytecodeGeneratorRunnable implements Runnable {
                         name = name.replaceAll("/", ".");
                         if (name.endsWith(".class")) {
                             try {
-                                src.processClasses(Class.forName(name.substring(0, name.length() - 6), false, amendedCL));
+                                src.processClasses(Class.forName(name.substring(0, name.length() - 6), false, loader));
                             } catch (ClassNotFoundException e) {
                                 throw new RuntimeException(e);
                             }
@@ -139,7 +144,7 @@ public class JmhBytecodeGeneratorRunnable implements Runnable {
             }
         } finally {
             try {
-                amendedCL.close();
+                loader.close();
             } catch (IOException e) {
                 e.printStackTrace();
             }

--- a/src/main/groovy/me/champeau/gradle/JmhBytecodeGeneratorTask.groovy
+++ b/src/main/groovy/me/champeau/gradle/JmhBytecodeGeneratorTask.groovy
@@ -34,7 +34,6 @@ class JmhBytecodeGeneratorTask extends DefaultTask {
     @OutputDirectory
     File generatedSourcesDir
 
-
     @Input
     String generatorType = 'default'
 
@@ -48,12 +47,12 @@ class JmhBytecodeGeneratorTask extends DefaultTask {
         def workerExecutor = getServices().get(WorkerExecutor)
         workerExecutor.submit(JmhBytecodeGeneratorRunnable) { WorkerConfiguration config ->
             config.isolationMode = IsolationMode.PROCESS
-            def classpath = runtimeClasspath.files
+            config.classpath = project.configurations.getByName("jmh");
+            def benchmarkClasspath = runtimeClasspath.files
             if (includeTests) {
-                classpath += testClasses.files + testRuntimeClasspath.files
+                benchmarkClasspath += testClasses.files + testRuntimeClasspath.files
             }
-            config.classpath = classpath
-            config.params(classesDirs.files as File[], generatedSourcesDir, generatedClassesDir, generatorType)
+            config.params(benchmarkClasspath as File[], classesDirs.files as File[], generatedSourcesDir, generatedClassesDir, generatorType)
         }
     }
 

